llama-index-core/llama_index/core/schema.py

```diff
@@ -389,7 +431,123 @@ def as_related_node_info(self) -> RelatedNodeInfo:
         )
 
 
+class MediaResource(BaseModel):
+    """A container class for media content.
+
+    This class represents a generic media resource that can be stored and accessed
+    in multiple ways - as raw bytes, on the filesystem, or via URL. It also supports
+    storing vector embeddings for the media content.
+
+    Attributes:
+        embeddings: Dense vector representation of this resource for embedding-based search/retrieval
+        sparse: Sparse vector representation of this resource
+        data: Raw binary data of the media content
+        mimetype: The MIME type indicating the format/type of the media content
+        path: Local filesystem path where the media content can be accessed
+        url: URL where the media content can be accessed remotely
+    """
+
+    embeddings: list[float] | None = Field(
+        default=None, description="Vector representation of this resource."
+    )
+    sparse: list[float] | None = Field(
+        default=None, description="Sparse vector representation of this resource."
+    )
+    data: bytes | None = Field(
+        default=None, exclude=True, description="Binary data of this resource."
+    )
+    mimetype: str | None = Field(
+        default="text/plain", description="MIME type of this resource."
+    )
+    path: Path | None = Field(
+        default=None, description="Filesystem path of this resource."
+    )
+    url: AnyUrl | None = Field(default=None, description="URL to reach this resource.")
+
+    @property
+    def hash(self) -> str:
+        """Generate a hash to uniquely identify the media resource.
+
+        The hash is generated based on the available content (data, path, or url).
+        Returns an empty string if no content is available.
+        """
+        if self.data is not None:
+            # Hash the binary data if available
+            return str(sha256(self.data).hexdigest())
+        elif self.path is not None:
+            # Hash the file contents if a path is provided
+            return str(sha256(self.path.read_bytes()).hexdigest())
+        elif self.url is not None:
+            # Use the URL string as basis for hash
+            return str(sha256(str(self.url).encode("utf-8")).hexdigest())
+        else:
+            # Return empty string if no content is available
+            return ""
+
+
+class Node(BaseNode):
```

Comment:
Just checking, but this should hopefully serialize fine? I know pydantic 2.0 has some weird default rules around serializing nested models. `node.model_dump()` should be enough to test, might be worth having a small unit test for this

Reply:
I think it would but I'll check! My only doubt is whether or not we want binary data into the model dump, is it a valid concern?

Reply:
Yea its a good point -- today, we dump it by default if its there 😓